d2common/d2datautils/stream_writer.go

```diff
@@ -28,6 +30,66 @@ func (v *StreamWriter) PushBytes(b ...byte) {
 	}
 }
 
+// PushBit pushes single bit into stream
+// WARNING: if you'll use PushBit, offset'll be less then 7, and if you'll
+// use another Push... method, bits'll not be pushed
+func (v *StreamWriter) PushBit(b bool) {
+	if b {
+		v.bitCache |= (1 << v.bitOffset)
+	}
+	v.bitOffset++
+
+	if v.bitOffset != bitsPerByte {
+		return
+	}
+
+	v.PushBytes(v.bitCache)
+	v.bitCache = 0
+	v.bitOffset = 0
+}
+
+// PushBits pushes bits (with max range 8)
+func (v *StreamWriter) PushBits(b byte, bits int) {
+	val := b
+	for i := 0; i < bits; i++ {
+		if val&1 == 1 {
+			v.PushBit(true)
+		} else {
+			v.PushBit(false)
+		}
+
+		val >>= 1
+	}
+}
+
+// PushBits16 pushes bits (with max range 16)
+func (v *StreamWriter) PushBits16(b uint16, bits int) {
+	val := b
+	for i := 0; i < bits; i++ {
+		if val&1 == 1 {
+			v.PushBit(true)
+		} else {
+			v.PushBit(false)
+		}
+
+		val >>= 1
+	}
+}
+
+// PushBits32 pushes bits (with max range 32)
+func (v *StreamWriter) PushBits32(b uint32, bits int) {
+	val := b
+	for i := 0; i < bits; i++ {
+		if val&1 == 1 {
+			v.PushBit(true)
+		} else {
+			v.PushBit(false)
+		}
+
+		val >>= 1
+	}
+}
+
```

Comment:
Same as previous comment; there is an assumed max size but this is not checked.

Reply:
I suppose since you are right-shifting this is fine, but maybe we should return if we know we've gone too far.

Reply:
also aadfbbc